CPDShell/Core/algorithms/knn_algorithm.py

```diff
@@ -0,0 +1,175 @@
+        """
+        Calculate the statistics of the KNN graph in specified point.
+
+        :param time: index of point in the given sample to calculate statistics relative to it.
+        :param window_size: size of sample to analyze.
+        """
+        assert self.__knngraph is not None, "Graph should not be None."
+
+        k = self.__k
+        n = window_size
+        n_1 = time
+        n_2 = n - time
+
+        if n <= k:
+            # Unable to analyze sample due to its size.
+            return -k
+
+        h = 4 * (n_1 - 1) * (n_2 - 1) / ((n - 2) * (n - 3))
+
+        sum_1 = (1 / n) * sum(
+            self.__knngraph.check_for_neighbourhood(i, j) * self.__knngraph.check_for_neighbourhood(j, i)
+            for i in range(window_size)
+            for j in range(window_size)
+        )
+
+        sum_2 = (1 / n) * sum(
+            self.__knngraph.check_for_neighbourhood(j, i) * self.__knngraph.check_for_neighbourhood(m, i)
+            for i in range(window_size)
+            for j in range(window_size)
+            for m in range(window_size)
+        )
+
+        expectation = 4 * k * n_1 * (n_2) / (n - 1)
+        variance = (expectation / k) * (h * (sum_1 + k - (2 * k**2 / (n - 1))) + (1 - h) * (sum_2 - k**2))
+        deviation = sqrt(variance)
+
+        permutation: np.array = np.arange(window_size)
+        # np.random.shuffle(permutation) # It seems that random permutation spoils the result
+
+        random_variable_value = self.__calculate_random_variable(permutation, time, window_size)
+
+        if deviation == 0:
+            # if deviation is zero, it likely means that time is 1. This implies that h is 0 and sum_2 = k**2.
+            # In this case we can for sure say that there is no change-point.
+            return -(random_variable_value - expectation)
+
+        statistics = -(random_variable_value - expectation) / deviation
+
+        return statistics
+
+    def __check_change_point(self, statistics: float) -> bool:
+        """
+        Check if calculated statistics is more than a given threshold to find out if it is a change point or not.
+
+        :return: True if change point occurs, False otherwise.
+        """
+        return statistics > self.__threshold
+
+    def __calculate_random_variable(self, permutation: np.array, t: int, window_size: int) -> int:
+        """
+        Calculate a random variable from a permutation and a fixed point.
+
+        :param permutation: random permutation of observations.
+        :param t: fixed point that splits the permutation.
+        :return: value of the random variable.
+        """
+
+        def b(i: int, j: int) -> bool:
+            pi = permutation[i]
+            pj = permutation[j]
+            return (pi <= t < pj) or (pj <= t < pi)
+
+        s = sum(
+            (self.__knngraph.check_for_neighbourhood(i, j) + self.__knngraph.check_for_neighbourhood(j, i)) * b(i, j)
```

Comment:
Couldn't we replace `*` with `and`, `+` with `or` here? Or clarify somehow (a comment) that we're summig here some booleans. It took me some time to understand a typing here, however, this is a valid code.

Reply:
We cannot replace `*` because `True + True == 2` and `True or True == True`.